intel-hex.js

```diff
@@ -648,57 +660,179 @@ function arraysToHex(blocks, lineSize = 16) {
-                    255                                  // Maximum record length as per spec
-                );
+            if (blockAddr < (highAddress + lowAddress)) {
+                throw new Error(
+                    'Block starting at 0x' +
+                    blockAddr.toString(16) +
+                    ' overlaps with a previous block.');
+            }
 
-                if (recordSize) {
+            lowAddress = blockAddr % 0x10000;
+            let blockOffset = 0;
+            let blockEnd = blockAddr + blockSize;
+            if (blockEnd > 0xFFFFFFFF) {
+                throw new Error('Data cannot be over 0xFFFFFFFF');
+            }
+
+            // Loop for every 64KiB memory segment that spans this block
+            while (highAddress + lowAddress < blockEnd) {
 
-                    recordHeader[0] = recordSize;   // Length
-                    recordHeader[1] = lowAddress >> 8;    // Load offset, high byte
-                    recordHeader[2] = lowAddress;    // Load offset, low byte
-                    recordHeader[3] = 0;    // Record type
+                if (lowAddress > 0xFFFF) {
+                    // Insert a new 0x04 record to jump to a new 64KiB block
+                    highAddress += 1 << 16; // Increase by one
+                    lowAddress = 0;
 
-                    let subBlock = block.subarray(blockOffset, blockOffset + recordSize);   // Data bytes for this record
+                    offsetRecord[0] = 2;    // Length
+                    offsetRecord[1] = 0;    // Load offset, high byte
+                    offsetRecord[2] = 0;    // Load offset, low byte
+                    offsetRecord[3] = 4;    // Record type
+                    offsetRecord[4] = highAddress >> 24;    // new address offset, high byte
+                    offsetRecord[5] = highAddress >> 16;    // new address offset, low byte
 
                     records.push(
                         ':' +
-                        Array.prototype.map.call(recordHeader, hexpad).join('') +
-                        Array.prototype.map.call(subBlock, hexpad).join('') +
-                        hexpad(checksumTwo(recordHeader, subBlock))
+                        Array.prototype.map.call(offsetRecord, hexpad).join('') +
+                        hexpad(checksum(offsetRecord))
                     );
+                }
+
+                let recordSize = -1;
+                // Loop for every record for that spans the current 64KiB memory segment
+                while (lowAddress < 0x10000 && recordSize) {
+                    recordSize = Math.min(
+                        lineSize,                            // Normal case
+                        blockEnd - highAddress - lowAddress, // End of block
+                        0x10000 - lowAddress,                // End of low addresses
+                        255                                  // Maximum record length as per spec
```

Comment:
If max record length is given by the spec, then perhaps vaildate that `lineSize` is not larger than max?